src/UglyToad.PdfPig/Writer/PdfStreamWriter.cs

```diff
@@ -0,0 +1,166 @@
+                throw new ArgumentNullException(nameof(catalogReference));
+            
+            WriteString($"%PDF-{version:0.0}", Stream);
+
+            Stream.WriteText("%");
+            Stream.WriteByte(169);
+            Stream.WriteByte(205);
+            Stream.WriteByte(196);
+            Stream.WriteByte(210);
+            Stream.WriteNewLine();
+
+            var offsets = new Dictionary<IndirectReference, long>();
+            ObjectToken catalogToken = null;
+            foreach(var pair in tokenReferences)
+            {
+                var referenceToken = pair.Key;
+                var token = pair.Value;
+                var offset = Stream.Position;
+                var obj = new ObjectToken(offset, referenceToken.Data, token);
+
+                TokenWriter.WriteToken(obj, Stream);
+
+                offsets.Add(referenceToken.Data, offset);
+
+                if (catalogToken == null && referenceToken == catalogReference)
+                {
+                    catalogToken = obj;
+                }
+            }
+
+            if (catalogToken == null)
+            {
+                throw new Exception("Catalog object wasn't found");
+            }
+
+            // TODO: Support document information
+            TokenWriter.WriteCrossReferenceTable(offsets, catalogToken, Stream, null);
+        }
+
+        public IndirectReferenceToken WriteToken(IToken token, int? reservedNumber = null)
+        {
+            // if you can't consider deduplicating the token. 
+            // It must be because it's referenced by his child element, so you must have reserved a number before hand
+            // Example /Pages Obj
+            var canBeDuplicated = !reservedNumber.HasValue;
+            if (!canBeDuplicated)
+            {
+                if (!reservedNumbers.Remove(reservedNumber.Value))
+                {
+                    throw new InvalidOperationException("You can't reuse a reserved number");
+                }
+
+                // When we end up writing this token, all of his child would already have been added and checked for duplicate
+                return AddToken(token, reservedNumber.Value);
+            }
+
+            var reference = FindToken(token);
+            if (reference == null)
+            {
+                return AddToken(token, CurrentNumber++);
+            }
+
+            return reference;
+        }
+        
+        public int ReserveNumber()
+        {
+            var reserved = CurrentNumber;
+            reservedNumbers.Add(reserved);
+            CurrentNumber++;
+            return reserved;
+        }
+
+        public IndirectReferenceToken ReserveNumberToken()
+        { 
+            return new IndirectReferenceToken(new IndirectReference(ReserveNumber(), 0));
+        }
+        
+        public byte[] ToArray()
+        {
+            if (!Stream.CanSeek)
+                throw new NotSupportedException($"{Stream.GetType()} can't seek");
```

Comment:
Should we let the `Stream` throw this exceptions? As of right know we are missing to check if we can read (`CanRead`)

Reply:
I alternate between letting the stream throw it and adding the check myself when I'm feeling eager. I'd add the CanRead check as well and move the check to the constructor in order to fail as early as possible if the stream is invalid, or remove it and let the stream fail when it gets used.

Reply:
I thought about the possibility of putting the checks in the constructor. But since is possible to use the Stream without ever calling this method, I leave them here. So for that reason I would let the inner stream throw